### PR TITLE
feat(dynamic-links): add support for utmParameters

### PR DESCRIPTION
**Pull request by Minishlink**
### Description

This adds the information about utm parameters that was setup in the dynamic link. This can be used to then store that information elsewhere or react to it somehow.

### Related issues

Might be useful for #4554

### Checklist

- I read the [Contributor Guide](../CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
  - [x] Yes
- My change supports the following platforms;
  - [x] `Android`
  - [x] `iOS`
- My change includes tests;
  - [x] `e2e` tests added or updated in `packages/\*\*/e2e`
  - [ ] `jest` tests added or updated in `packages/\*\*/__tests__`
- [x] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [ ] Yes
  - [x] No

### Test Plan

---

Think `react-native-firebase` is great? Please consider supporting the project with any of the below:

- 👉 Star this repo on GitHub ⭐️
- 👉 Follow [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter

:fire:

### CODE REVIEW

File: packages/dynamic-links/e2e/dynamicLinks.e2e.js [hunk 1]
```diff
@@ -155,6 +155,7 @@ describe('dynamicLinks()', function () {
 
       dynamicLink.should.be.an.Object();
       dynamicLink.url.should.equal('https://rnfirebase.io');
+      dynamicLink.utmParameters.should.eql({});
```
Comment (mikehardy) [inline packages/dynamic-links/e2e/dynamicLinks.e2e.js:158]: With a quick addition of utm params to TEST_LINK the population may be verified I think?

File: packages/dynamic-links/e2e/dynamicLinks.e2e.js [hunk 2]
```diff
@@ -172,6 +173,7 @@ describe('dynamicLinks()', function () {
 
       spy.getCall(0).args[0].should.be.an.Object();
       spy.getCall(0).args[0].url.should.equal('https://invertase.io/hire-us');
+      spy.getCall(0).args[0].utmParameters.should.eql({});
```
Comment (mikehardy) [inline packages/dynamic-links/e2e/dynamicLinks.e2e.js:176]: same here - quick addition of utm params to TEST_LINK2 and this may be verified

**Review by mikehardy — CHANGES_REQUESTED** [review]
Oh this is really cool! Thank you for posting this, it should be really useful.
firebase-js-sdk does not have a links module to compare types with and it seems to me `Record<string, string>` is as good as any

My only hesitation - especially since I think it's an easy addition? - is that it could be verified much more strongly in the e2e test

I scanned the test carefully and left two specific notes for the initial / onlink test, and then for the resolve link if you added the params at the top of this one (where minimum version was added) and verified it at the bottom (where minimum version was checked) I think all the work here would be fully verified

https://github.com/invertase/react-native-firebase/pull/5593/files#diff-b885fc4a786a779f37b59dd276f88b9625d573aee097b53e5d8b4f47789a306aR80-R101

The code itself in Java/Objective-C looks great but I remember when I added resolveLink my code also looked great to me then I added verification in e2e and of course there was some subtle problem :laughing: 

If you could try that and see if it really verifies in the e2e test I'd appreciate it and could merge+release straight away

**Comment by mikehardy** [comment]
There weren't a lot of PRs in flight in the repo so I chose this moment to integrate an auto formatter (#5596) - which necessarily cause all sorts of conflicts in this PR. That was no fault of the PR's of course, so I merged it and I believe I have correctly resolved all the conflicts with the last commit here if you pull it locally

File: packages/dynamic-links/ios/RNFBDynamicLinks/RNFBDynamicLinksAppDelegateInterceptor.m [hunk 3]
```diff
@@ -99,7 +105,10 @@ - (BOOL)application:(UIApplication *)application
                          @"minimumAppVersion" : dynamicLink.minimumAppVersion == nil
                              ? [NSNull null]
                              : dynamicLink.minimumAppVersion,
-                       }];
+                         @"utmParameters": dynamicLink.utmParametersDictionary == nil
+                             ? @{}
+                             : dynamicLink.utmParametersDictionary,
+                         }];
```
Comment (mikehardy) [inline packages/dynamic-links/ios/RNFBDynamicLinks/RNFBDynamicLinksAppDelegateInterceptor.m:111]: To my close examination this looks like a very minor formatting error vs the original post-auto-format
this should fix it
```suggestion
                       }];
```

File: packages/dynamic-links/android/src/main/java/io/invertase/firebase/dynamiclinks/ReactNativeFirebaseDynamicLinksModule.java [hunk 4]
```diff
@@ -435,7 +441,8 @@ public void onNewIntent(Intent intent) {
                               "dynamic_links_link_received",
                               dynamicLinkToWritableMap(
                                   pendingDynamicLinkData.getLink().toString(),
-                                  pendingDynamicLinkData.getMinimumAppVersion())));
+                                  pendingDynamicLinkData.getMinimumAppVersion(),
+                                  Arguments.makeNativeMap(pendingDynamicLinkData.getUtmParameters()));
```
Comment (mikehardy) [inline packages/dynamic-links/android/src/main/java/io/invertase/firebase/dynamiclinks/ReactNativeFirebaseDynamicLinksModule.java:445]: formatting fallout, on me to fix
```suggestion
                                  Arguments.makeNativeMap(pendingDynamicLinkData.getUtmParameters())));
```

File: packages/dynamic-links/android/src/main/java/io/invertase/firebase/dynamiclinks/ReactNativeFirebaseDynamicLinksModule.java [hunk 5]
```diff
@@ -435,7 +441,8 @@ public void onNewIntent(Intent intent) {
                               "dynamic_links_link_received",
                               dynamicLinkToWritableMap(
                                   pendingDynamicLinkData.getLink().toString(),
-                                  pendingDynamicLinkData.getMinimumAppVersion())));
+                                  pendingDynamicLinkData.getMinimumAppVersion(),
+                                  Arguments.makeNativeMap(pendingDynamicLinkData.getUtmParameters())));
```
Comment (mikehardy) [inline packages/dynamic-links/android/src/main/java/io/invertase/firebase/dynamiclinks/ReactNativeFirebaseDynamicLinksModule.java:445]: I can count parentheses, promise :crossed_fingers: 
```suggestion
                                  Arguments.makeNativeMap(pendingDynamicLinkData.getUtmParameters()))));
```

**Comment by Minishlink** [comment]
Unfortunately it is not possible to test these cases (or am I missing something?), as apparently Firebase does not rely on the long link and does not parse utm parameters on a short link, it only works with a short link that has utm parameters inside. This should be more understandable:

- `https://reactnativefirebase.page.link/?link=https://invertase.io/hire-us&apn=com.invertase.testing&utm_source=github` : you won't get any utm parameters from getUtmParameters
- `https://reactnativefirebase.page.link/shortLinkWithoutUtmParameters?utm_source=github` : you won't get any utm parameters from getUtmParameters
- `https://reactnativefirebase.page.link/shortLinkWithUtmParameters` : you will get the utm parameters. These utm parameters corresponds to the utm parameters you put when generating that short link in the Firebase dashboard (or through the in app builder I guess)

**Comment by mikehardy** [comment]
OK I see what you mean. However, it's possible to generate a short link with utm params, then resolve it isn't it? That's what I meant to suggest. Worst case I have access to the dashboard for the testing project and can make a link with utm params in it, I was just hoping it was testable with completely self contained data fixture

**Comment by mikehardy** [comment]
I had some time to check this while waiting for some other tests to run, if you use buildShortLink in the e2e test then the params do come through - I've got it locally, I'll push to your branch then merge this if I can, or if I mess up the git magic somehow I'll carry the commit in a fresh PR. Either way, all of your native parsing and everything is :100: - it all checked out first time when I ran it here locally

**Review by mikehardy — APPROVED** [review]
the changes needed to do the test with UTM params in it exposed a couple other problems, and fixing those on this branch + in this PR isn't good practice. This appears to be working well in local testing so I'm going to merge it then dig in to the other issues in the area.

Thanks again for posting this!

**Pull request merged**